lib/lower/convert-expr.cc

```diff
@@ -0,0 +1,921 @@
+    return {};
+  }
+
+  /// Generate an integral constant of `value`
+  template <int KIND>
+  M::Value genIntegerConstant(M::MLIRContext *context, std::int64_t value) {
+    M::Type type{converter.genType(IntegerCat, KIND)};
+    auto attr{builder.getIntegerAttr(type, value)};
+    auto res{builder.create<M::ConstantOp>(getLoc(), type, attr)};
+    return res.getResult();
+  }
+
+  /// Generate a logical/boolean constant of `value`
+  M::Value genLogicalConstantAsI1(M::MLIRContext *context, bool value) {
+    M::Type i1Type{M::IntegerType::get(1, builder.getContext())};
+    auto attr{builder.getIntegerAttr(i1Type, value ? 1 : 0)};
+    return builder.create<M::ConstantOp>(getLoc(), i1Type, attr).getResult();
+  }
+
+  template <int KIND>
+  M::Value genRealConstant(M::MLIRContext *context, L::APFloat const &value) {
+    M::Type fltTy{convertReal(context, KIND)};
+    auto attr{builder.getFloatAttr(fltTy, value)};
+    auto res{builder.create<M::ConstantOp>(getLoc(), fltTy, attr)};
+    return res.getResult();
+  }
+
+  M::Type getSomeKindInteger() {
+    return M::IndexType::get(builder.getContext());
+  }
+
+  template <typename OpTy, typename A>
+  M::Value createBinaryOp(A const &ex, M::Value lhs, M::Value rhs) {
+    assert(lhs && rhs && "argument did not lower");
+    auto x = builder.create<OpTy>(getLoc(), lhs, rhs);
+    return x.getResult();
+  }
+  template <typename OpTy, typename A>
+  M::Value createBinaryOp(A const &ex, M::Value rhs) {
+    return createBinaryOp<OpTy>(ex, genval(ex.left()), rhs);
+  }
+  template <typename OpTy, typename A>
+  M::Value createBinaryOp(A const &ex) {
+    return createBinaryOp<OpTy>(ex, genval(ex.left()), genval(ex.right()));
+  }
+
+  M::FuncOp getFunction(L::StringRef name, M::FunctionType funTy) {
+    auto module{getModule(&builder)};
+    if (M::FuncOp func{getNamedFunction(module, name)}) {
+      assert(func.getType() == funTy &&
+             "function already declared with a different type");
+      return func;
+    }
+    return createFunction(module, name, funTy);
+  }
+
+  M::FuncOp getRuntimeFunction(RuntimeEntryCode callee, M::FunctionType funTy) {
+    auto name{getRuntimeEntryName(callee)};
+    return getFunction(name, funTy);
+  }
+
+  // FIXME binary operation :: ('a, 'a) -> 'a
+  template <Co::TypeCategory TC, int KIND>
+  M::FunctionType createFunctionType() {
+    if constexpr (TC == IntegerCat) {
+      M::Type output{converter.genType(IntegerCat, KIND)};
+      L::SmallVector<M::Type, 2> inputs;
+      inputs.push_back(output);
+      inputs.push_back(output);
+      return M::FunctionType::get(inputs, output, builder.getContext());
+    } else if constexpr (TC == RealCat) {
+      M::Type output{convertReal(builder.getContext(), KIND)};
+      L::SmallVector<M::Type, 2> inputs;
+      inputs.push_back(output);
+      inputs.push_back(output);
+      return M::FunctionType::get(inputs, output, builder.getContext());
+    } else {
+      assert(false);
+      return {};
```

Comment:
Why the {}? after return?

Reply:
It means the same thing as `return mlir::FunctionType();` here.